### PR TITLE
fix missing sha256sum command and print info for fish shell

### PR DESCRIPTION
**Pull request by khaledez**
When I tried to install ponyup on my macOS machine, it was failing with this error:
```bash
sh: line 116: sha256sum: command not found
```

But  macOS Catalina comes with `shasum` utility, so I made a change to check if `sha256sum` is not available, then use `shasum`, otherwise just exit printing error.

Also, I use [fish shell](https://fish.sh) which has a slightly different syntax, so I added a custom message for fish.

### CODE REVIEW

File: ponyup-init.sh [hunk 1]
```diff
@@ -27,6 +27,17 @@ BLUE="\e[34m"
 RED="\e[31m"
 YELLOW="\e[33m"
 
+shasumCommand() {
+  if command -v sha256sum > /dev/null 2>&1; then
+    sha256sum "$@"
+  elif command -v shasum > /dev/null 2>&1; then
+    shasum --algorithm 256 "$@"
+  else 
+    echo "No checksum command!"
```
Comment (SeanTAllen) [inline ponyup-init.sh]: can you update this to print in RED and use printf instead of echo like:

`printf "%bNo checksum command found.%b\n"` "${RED}" "${DEFAULT}"`

**Comment by SeanTAllen** [comment]
Can you also add 2 entries to the CHANGELOG as part of this PR?

- One for FIXED related to the shasum issue.
- One for ADDED for the fish shell specific message

File: ponyup-init.sh [hunk 2]
```diff
@@ -2,6 +2,7 @@
 
 set -o errexit
 set -o nounset
+set -o pipefail
```
Comment (SeanTAllen) [inline ponyup-init.sh]: This isn't a valid option for Posix shell support. We are keeping this file working with sh support. You'll need to remove. You'll see CI failed after the addition of this.

File: ponyup-init.sh [hunk 3]
```diff
@@ -27,6 +28,17 @@ BLUE="\e[34m"
 RED="\e[31m"
 YELLOW="\e[33m"
 
+shasumCommand() {
+  if command -v sha256sum > /dev/null 2>&1; then
+    sha256sum "$@"
+  elif command -v shasum > /dev/null 2>&1; then
+    shasum --algorithm 256 "$@"
+  else 
+    printf "%bNo checksum command found.%b\n" "${RED}" "${DEFAULT}" >&2
```
Comment (SeanTAllen) [inline ponyup-init.sh]: No `>&2` is needed for how we are currently doing error output.

If you want, you can open an issue to address error message not going to STDERR. I'd want that to be part of a separate PR though.

File: ponyup-init.sh [hunk 4]
```diff
@@ -130,10 +142,10 @@ echo "downloading ${filename}"
 
 curl "${dl_url}" -o "${tmp_dir}/${filename}"
 
-dl_checksum="$(sha256sum "${tmp_dir}/${filename}" | awk '{ print $1 }')"
+dl_checksum="$(shasumCommand "${tmp_dir}/${filename}" | awk '{ print $1 }')"
 
 if [ "${dl_checksum}" != "${checksum}" ]; then
-  printf "%bmchecksum mismatch:\n" "${RED}"
+  printf "%bchecksum mismatch:\n" "${RED}"
```
Comment (SeanTAllen) [inline ponyup-init.sh]: This is an unrelated change. Can you remove and we can get in another PR?

Comment (SeanTAllen) [reply ponyup-init.sh]: I'll push that change now. Thanks for spotting.

Comment (SeanTAllen) [reply ponyup-init.sh]: I've opened a PR for this.

https://github.com/ponylang/ponyup/pull/93

File: ponyup-init.sh (hunk 3 above)
Comment (khaledez) [reply ponyup-init.sh]: I made it like this to ensure printing the error message, otherwise it will look like this:
Screen Shot 2020-02-04 at 9 24 04 PM

But with `pipefail` and redirecting to `stderr` it looks like this:
Screen Shot 2020-02-04 at 9 25 03 PM

Do you know how to solve it and comply with posix?

Comment (SeanTAllen) [reply ponyup-init.sh]: Not off the top of my head. Can you open an issue for that? It's different than what you are currently tackling so best to do in its own issue and PR.

Comment (khaledez) [reply ponyup-init.sh]: So, it's ok for this PR to have the wrong error message?

Comment (SeanTAllen) [reply ponyup-init.sh]: Wait, so...

without the redirect it isn't exiting correctly?

Comment (khaledez) [reply ponyup-init.sh]: it's exiting, but the error message will be misleading! like the first screenshot!

Comment (SeanTAllen) [reply ponyup-init.sh]: Ok, well then, the shasum call will need to be handled in a more compliant fashion that doesn't require the pipefail. I'm not sure what that would be off the top of my head. You'll need to investigate that.

Sorry, I didn't understand what the issue was until now. Apologies.

Comment (khaledez) [reply ponyup-init.sh]: No worries, I'll investigate it and see what can I do

Comment (SeanTAllen) [reply ponyup-init.sh]: It exits as a result of the checksum failing. The exit in the function doesn't appear to be working. I'm not sure why and don't have time at the moment to look, you'll need to figure out why it isn't an address in a posix compliant fashion.

My suggestion would be near the start of the script, test for a valid shasum and set a variable to that, then use the variable rather than a function call. And you can exit at the top.

Does that make sense? I hope I was clear.

I'd say do that check and variable set up before:

`ponyup_root="${prefix}/ponyup"`

Comment (khaledez) [reply ponyup-init.sh]: Makes sense, that's what I was thinking too.

Comment (khaledez) [reply ponyup-init.sh]: Done

File: ponyup-init.sh [hunk 5]
```diff
@@ -130,7 +139,7 @@ echo "downloading ${filename}"
 
 curl "${dl_url}" -o "${tmp_dir}/${filename}"
 
-dl_checksum="$(sha256sum "${tmp_dir}/${filename}" | awk '{ print $1 }')"
+dl_checksum="$($sha256sum "${tmp_dir}/${filename}" | awk '{ print $1 }')"
```
Comment (SeanTAllen) [inline ponyup-init.sh]: one more small change. can you do

${sha256sum} rather than $sha256sum

**Comment by SeanTAllen** [comment]
Thank you very much for this @khaledez and welcome to folks who have contributed to the project.

**Comment by khaledez** [comment]
Thank you @SeanTAllen for your help and quick response!
This is officially my first open source contribution 😄

**Comment by SeanTAllen** [comment]
@khaledez you are welcome. FYI, Your change is going out in a release RIGHT NOW. So, how is that for instant gratification?

**Comment by khaledez** [comment]
Amazing!
